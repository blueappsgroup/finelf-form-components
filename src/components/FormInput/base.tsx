/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { ReactElement, useContext, ChangeEvent } from 'react'
import styled from 'styled-components'
import { FieldWrapProps, SelectFieldWrapProps } from '../../types'
import { device } from '../../consts/sizes'
import { FormContext, setFormValuesToCache } from '../../utils'

type StyledProps = {
  theme: {
    inputErrorColor: string
    inputBgColor: string
    inputPlaceHolderColor: string
    inputBorderColor: string
    inputTextColor: string
  }
  error?: string | boolean
}

const StyledRow = styled.div`
  display: flex;
  align-items: center;
  margin-bottom: 12px;
  position: relative;
`

export const Row = styled.div`
  display: flex;
  flex-direction: column;

  @media ${device.tablet} {
    flex-direction: row;
    & ${StyledRow}:first-of-type {
      flex-grow: 1;
      margin-right: 5px;
    }

    & ${StyledRow}:last-child {
      flex-grow: 1;
      margin-left: 5px;
    }
  }
`

const StyledError = styled.span`
  color: ${(props: StyledProps): string => props.theme.inputErrorColor};
  position: absolute;
  font-size: 10px;
  bottom: -12px;
`

const StyledInput = styled.input<any>`
  background: ${(props: StyledProps): string => props.theme.inputBgColor};
  border: 1px solid
    ${(props: StyledProps): string => props.theme.inputBorderColor};
  box-sizing: border-box;
  box-shadow: 0px 7px 64px rgba(194, 186, 186, 0.07);
  border-radius: 6px;
  width: 100%;
  height: 63px;
  display: flex;
  align-items: center;
  font-style: normal;
  font-weight: 500;
  font-size: 14px;
  line-height: 24px;
  padding: 0px 20px;
  border-color: ${(props: StyledProps): string =>
    props.error ? props.theme.inputErrorColor : props.theme.inputBorderColor};
  color: ${(props: any): string =>
    props.error ? props.theme.inputErrorColor : props.theme.inputTextColor};

  &::placeholder {
    color: ${(props: StyledProps): string => props.theme.inputPlaceHolderColor};
  }

  &:focus {
    outline: none;
  }
`

const StyledSelect = styled.select<any>`
  background: ${(props: StyledProps): string => props.theme.inputBgColor};
  border: 1px solid
    ${(props: StyledProps): string => props.theme.inputBorderColor};
  box-sizing: border-box;
  box-shadow: 0px 7px 64px rgba(194, 186, 186, 0.07);
  border-radius: 6px;
  width: 100%;
  height: 63px;
  display: flex;
  align-items: center;
  font-style: normal;
  font-weight: 500;
  font-size: 14px;
  line-height: 24px;
  padding: 0px 20px;
  border-color: ${(props: StyledProps): string =>
    props.error ? props.theme.inputErrorColor : props.theme.inputBorderColor};
  color: ${(props: any): string =>
    props.error ? props.theme.inputErrorColor : props.theme.inputTextColor};

  &::placeholder {
    color: ${(props: StyledProps): string => props.theme.inputPlaceHolderColor};
  }

  &:focus {
    outline: none;
  }
`

const BaseField: (props: FieldWrapProps) => ReactElement = ({
  field,
  form: { touched, errors, values },
  ...props
}) => {
  const { id } = useContext(FormContext)
  const handleOnBlur = (e: ChangeEvent<HTMLInputElement>): void => {
    setFormValuesToCache(values, id)
    field.onBlur && field.onBlur(e)
  }

  return (
    <StyledRow>
      {props.label && <label htmlFor={field.name}>{props.label}</label>}
      <StyledInput
        {...field}
        {...props}
        onBlur={handleOnBlur}
        type={props.type}
        value={(field.value && field.value) || ''}
        error={touched[field.name] && errors[field.name]}
        placeholder={
          props.placeholder &&
          `${props.placeholder}${(props.required && '*') || ''}`
        }
      />
      {props.showError && touched[field.name] && errors[field.name] && (
        <StyledError>{errors[field.name]}</StyledError>
      )}
    </StyledRow>
  )
}

export default BaseField

export const BaseSelectField: (props: SelectFieldWrapProps) => ReactElement = ({
  field,
  form: { touched, errors, values },
  ...props
}) => {
  const { id } = useContext(FormContext)
  const handleOnBlur = (e: ChangeEvent<HTMLInputElement>): void => {
    setFormValuesToCache(values, id)
    field.onBlur && field.onBlur(e)
  }

  const options = (options: []): Array<JSX.Element> => {
    const opts = []

    for (const option in options) {
      opts.push(
        <option key={option} value={option}>
          {options[option]}
        </option>
      )
    }

    return opts
  }

  return (
    <StyledRow>
      {props.label && <label htmlFor={field.name}>{props.label}</label>}
      <StyledSelect
        {...field}
        {...props}
        onBlur={handleOnBlur}
        type="text"
        value={(field.value && field.value) || ''}
        error={touched[field.name] && errors[field.name]}
        placeholder={
          props.placeholder &&
          `${props.placeholder}${(props.required && '*') || ''}`
        }
      >
        {props.options && options(props.options)}
      </StyledSelect>
      {props.showError && touched[field.name] && errors[field.name] && (
        <StyledError>{errors[field.name]}</StyledError>
      )}
    </StyledRow>
  )
}
