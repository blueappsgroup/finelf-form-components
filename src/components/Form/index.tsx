import React, { FC, ReactElement, useMemo, useState } from 'react'
import { Formik, Form } from 'formik'
import styled from 'styled-components'

import { ThemeProvider } from '../../consts/theme'
import { device } from '../../consts/sizes'
import { FormProps, FormValuesType } from '../../types'
import RedirectPage from '../RedirectPage'
import {
  FormContext,
  getFormValuesFromCache,
  resetFormValueCache,
  handleSendDataToApi,
  getFieldsValuesFromUrl,
} from '../../utils'
import { formStatuses } from '../../consts/form'

const StyledForm = styled(Form)`
  display: flex;
  flex-direction: column;
  max-width: ${(props): string => props.theme.formMaxWidth};
  justify-self: center;
  background: ${(props): string => props.theme.formBgColor};
  font-family: ${(props): string => props.theme.fontFamily};
  border-radius: 6px;
  box-shadow: ${(props): string => props.theme.formBoxShadow};
`

const FormWrapper: FC<FormProps> = ({
  children,
  onSubmit,
  customTheme,
  id,
  stepsLength,
  stepsTitles,
  hasRedirect,
  redirectUrl,
  timeToRedirect,
  redirectHeaderText,
  logoImg,
  redirectMainImg,
  redirectBgImg,
  sendDataToApi,
  apiUrl,
  transactionName,
  propertyNamesFromUrl,
  dataWithUserAgent,
}) => {
  const trasationIdValue =
    transactionName &&
    new URLSearchParams(window.location.search).get(transactionName)
  const intialValuesFromUrl = useMemo(
    () =>
      (propertyNamesFromUrl &&
        propertyNamesFromUrl.length > 0 &&
        getFieldsValuesFromUrl(propertyNamesFromUrl)) ||
      {},
    [propertyNamesFromUrl]
  )
  const [redirectUrlPath, setRedirectUrlPath] = useState(redirectUrl)
  const [initialValues, setInitialValues] = useState({
    ...intialValuesFromUrl,
    ...getFormValuesFromCache(id),
    // eslint-disable-next-line @typescript-eslint/camelcase
    trasaction_id: trasationIdValue,
  })
  const [currentStep, setCurrentStep] = useState(0)
  const [fieldsForSkip, setFieldsForSkip] = useState<string[]>([])
  const [errorFromApi, setErrorFromApi] = useState<boolean>(false)
  const shouldRedirect = !errorFromApi && hasRedirect
  const addFieldForSkip = (key: string): void =>
    setFieldsForSkip([...fieldsForSkip, key])

  const handleSubmit = async (
    values: FormValuesType,
    props: {
      resetForm: () => void
      setStatus: (status: string) => void
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): Promise<any> => {
    if (onSubmit) {
      onSubmit(values, props)
    }

    try {
      if (sendDataToApi && apiUrl) {
        const response = await handleSendDataToApi(
          values,
          apiUrl,
          id,
          fieldsForSkip,
          dataWithUserAgent
        )
        const {
          redirectUrl: urlFromApi,
          status: statusFromApi,
        } = await response.json()

        if (statusFromApi === false) {
          setErrorFromApi(true)
          return
        }

        urlFromApi && setRedirectUrlPath(urlFromApi)
      }

      props.resetForm()
      props.setStatus(formStatuses.submited)
    } catch (e) {
      props.setStatus(formStatuses.error)
    }
  }
  const handleReset = (): void => {
    resetFormValueCache(id)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    setInitialValues({} as any)
  }

  const prevStep: Function = () => setCurrentStep(currentStep - 1)

  const nextStep: Function = () => setCurrentStep(currentStep + 1)

  return (
    <FormContext.Provider
      value={{
        id,
        apiUrl,
        stepsLength,
        currentStep,
        stepsTitleList: stepsTitles,
        nextStep,
        prevStep,
        fieldsForSkip,
        addFieldForSkip,
        errorFromApi,
      }}
    >
      <ThemeProvider customTheme={{ ...customTheme }}>
        <Formik
          enableReinitialize
          initialValues={initialValues}
          onSubmit={handleSubmit}
          onReset={handleReset}
        >
          {(props): ReactElement =>
            (shouldRedirect && props.status === formStatuses.submited && (
              <RedirectPage
                redirectUrl={redirectUrlPath}
                backgroundImage={redirectBgImg}
                logoImg={logoImg}
                headerText={redirectHeaderText}
                timeToRedirect={timeToRedirect}
                mainImg={redirectMainImg}
              />
            )) || (
              <StyledForm id={id}>
                {children}
              </StyledForm>
            )
          }
        </Formik>
      </ThemeProvider>
    </FormContext.Provider>
  )
}

export default FormWrapper
