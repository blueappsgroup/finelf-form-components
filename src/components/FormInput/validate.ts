import {
  isEmpty,
  isNotValidLength,
  isNotValidPhoneNumber,
  isNotValidEmail,
  isNotLetter,
} from './validateHelpers'

export const validateText: (
  required: boolean
) => (value: string) => string | undefined = (required: boolean) => (
  value: string
): string | undefined => {
  if (required && isEmpty(value)) {
    return 'To pole jest wymagane'
  }

  if (isNotValidLength(value)) {
    return 'Pole musi zawierać między 2 - 100 znaków'
  }

  if (isNotLetter(value)) {
    return 'Podane dane są nieprawidłowe'
  }
}

export const validateEmail: (
  customErrorMsg: string | undefined,
  required: boolean
) => (value: string) => string | undefined = (
  customErrorMsg = 'Podany email jest nieprawidłowy',
  required: boolean
) => (value: string): string | undefined => {
  if (required && isEmpty(value)) {
    return 'To pole jest wymagane'
  }

  if (isNotValidEmail(value)) {
    return customErrorMsg
  }
}

export const validateTelNumber: (
  customErrorMsg: string | undefined,
  required: boolean
) => (value: string) => string | undefined = (
  customErrorMsg = 'Podany numer telefonu jest nieprawidłowy',
  required: boolean
) => (value: string): string | undefined => {
  if (required && isEmpty(value)) {
    return 'To pole jest wymagane'
  }

  if (isNotValidPhoneNumber(value)) {
    return customErrorMsg
  }
}

export const validateCheckbox: (
  required: boolean
) => (value: string) => string | undefined = (required: boolean) => (
  value: string
): string | undefined => (required && !value && 'required') || undefined
