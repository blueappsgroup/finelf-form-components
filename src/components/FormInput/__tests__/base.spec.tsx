import * as React from 'react'
import { render, fireEvent, act } from '@testing-library/react'
import { Field } from 'formik'

import Form from '../../Form'
import { BaseSelectField, BaseRangeField } from '../base'
import BaseField from '../base'
import { validateText } from '../validate'

describe('base <BaseField />', () => {
  const onSubmit = jest.fn()
  const setupWrapper = ({
    formId = 'testForm',
    inputId = 'baseField',
    inputName = 'baseField',
    required = false,
    component = BaseField,
    ...rest
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  }): any => {
    const wrapper = render(
      <Form id={formId} onSubmit={onSubmit}>
        <Field
          id={inputId}
          name={inputName}
          component={component}
          validate={validateText(required)}
          required={required}
          {...rest}
        />
      </Form>
    )
    const { container } = wrapper

    const input = container.querySelector(`[name="${inputName}"]`)

    return {
      input,
      ...wrapper,
    }
  }

  it('matches snapshot', () => {
    const { input } = setupWrapper({})

    expect(input).toBeTruthy()
    expect(input).toMatchSnapshot()
  })

  it('matches snapshot with label', () => {
    const wrapper = setupWrapper({ label: 'Test' })
    const { container } = wrapper

    expect(wrapper.getByText('Test')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('matches snapshot with error', async () => {
    const wrapper = setupWrapper({
      label: 'Test',
      showError: true,
      required: true,
      error: 'true',
    })
    const { container, input } = wrapper

    await act(async () => {
      fireEvent.focus(input, { target: { value: 'test' } })
    })

    await act(async () => {
      fireEvent.blur(input, { target: { value: 'test' } })
    })

    expect(wrapper.getByText('To pole jest wymagane')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('changes input field', async () => {
    const { input } = setupWrapper({})

    await act(async () => {
      fireEvent.change(input, { target: { value: 'test' } })
    })

    expect(input.value).toBe('test')
  })

  it('input field onBlur save value to sessionStorage', async () => {
    const { input } = setupWrapper({})

    await act(async () => {
      fireEvent.change(input, { target: { value: 'test' } })
    })

    await act(async () => {
      fireEvent.blur(input, { target: { value: 'test' } })
    })

    expect(
      JSON.parse(global.window.sessionStorage.getItem('form-testForm'))
        .baseField
    ).toBe('test')
    expect(input.value).toBe('test')
  })
})

describe('base <BaseSelectField />', () => {
  const onSubmit = jest.fn()
  const setupWrapper = ({
    formId = 'testForm',
    inputId = 'baseSelectField',
    inputName = 'baseSelectField',
    required = false,
    component = BaseSelectField,
    options = { test1: 'test1', test2: 'test2' },
    ...rest
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  }): any => {
    const wrapper = render(
      <Form id={formId} onSubmit={onSubmit}>
        <Field
          id={inputId}
          name={inputName}
          component={component}
          validate={validateText(required)}
          required={required}
          options={options}
          {...rest}
        />
      </Form>
    )
    const { container } = wrapper

    const input = container.querySelector(`[name="${inputName}"]`)

    return {
      input,
      ...wrapper,
    }
  }

  it('matches snapshot', () => {
    const { input } = setupWrapper({})

    expect(input).toBeTruthy()
    expect(input).toMatchSnapshot()
  })

  it('matches snapshot with label', () => {
    const wrapper = setupWrapper({ label: 'Test' })
    const { container } = wrapper

    expect(wrapper.getByText('Test')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('changes input field', async () => {
    const { input } = setupWrapper({})

    await act(async () => {
      fireEvent.change(input, { target: { value: 'test1' } })
    })

    expect(input.value).toBe('test1')
  })

  it('input field onBlur save value to sessionStorage', async () => {
    const { input } = setupWrapper({})

    await act(async () => {
      fireEvent.change(input, { target: { value: 'test1' } })
    })

    await act(async () => {
      fireEvent.blur(input, { target: { value: 'test1' } })
    })

    expect(
      JSON.parse(global.window.sessionStorage.getItem('form-testForm'))
        .baseField
    ).toBe('test')
    expect(input.value).toBe('test1')
  })
})

describe('base <BaseRangeField />', () => {
  const onSubmit = jest.fn()
  const setupWrapper = ({
    formId = 'testForm',
    inputId = 'baseRangeField',
    inputName = 'baseRangeField',
    required = false,
    component = BaseRangeField,
    min = 300,
    max = 10000,
    step = 100,
    value = 2000,
    suffix = 'PLN',
    unit = 'zÅ‚',
    ...rest
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  }): any => {
    const wrapper = render(
      <Form id={formId} onSubmit={onSubmit}>
        <Field
          id={inputId}
          name={inputName}
          component={component}
          validate={validateText(required)}
          required={required}
          min={min}
          max={max}
          step={step}
          value={value}
          suffix={suffix}
          unit={unit}
          {...rest}
        />
      </Form>
    )
    const { container } = wrapper

    const input = container.querySelector(`[name="${inputName}"]`)
    const slider = container.querySelector('.rangeslider')

    return {
      input,
      slider,
      ...wrapper,
    }
  }

  it('matches snapshot', () => {
    const { input } = setupWrapper({})

    expect(input).toBeTruthy()
    expect(input).toMatchSnapshot()
  })

  it('matches snapshot with label', () => {
    const wrapper = setupWrapper({ label: 'Test' })
    const { container } = wrapper

    expect(wrapper.getByText('Test')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('changes input field', async () => {
    const { input } = setupWrapper({})

    await act(async () => {
      fireEvent.change(input, { target: { value: '400' } })
    })

    expect(input.value).toBe('400')
  })

  it('input field onBlur save value to sessionStorage', async () => {
    const { input } = setupWrapper({})

    await act(async () => {
      fireEvent.change(input, { target: { value: '400' } })
    })

    await act(async () => {
      fireEvent.blur(input, { target: { value: '400' } })
    })

    expect(
      JSON.parse(global.window.sessionStorage.getItem('form-testForm'))
        .baseField
    ).toBe('test')
    expect(input.value).toBe('400')
  })
})
