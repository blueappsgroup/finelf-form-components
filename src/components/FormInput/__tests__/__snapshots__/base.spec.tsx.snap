// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base <BaseField /> matches snapshot 1`] = `
<input
  class="sc-fodVxV kubVCF"
  id="baseField"
  name="baseField"
  value=""
/>
`;

exports[`base <BaseField /> matches snapshot with error 1`] = `
<div>
  <form
    action="#"
    class="sc-fubCfw clDFWv"
    id="testForm"
  >
    <div
      class="sc-iqHYGH jWqqni"
    >
      <label
        for="baseField"
      >
        Test*
      </label>
      <div
        class="sc-iJuUWI cFUPqY"
      >
        <input
          class="sc-fodVxV gZGAuo"
          id="baseField"
          label="Test"
          name="baseField"
          required=""
          value=""
        />
      </div>
      <span
        class="sc-hBEYos kIOtFs"
      >
        To pole jest wymagane
      </span>
    </div>
    <button
      class="sc-iBPRYJ hcYpVh"
      disabled=""
      type="submit"
    >
      Wyślij
    </button>
  </form>
</div>
`;

exports[`base <BaseField /> matches snapshot with label 1`] = `
<div>
  <form
    action="#"
    class="sc-fubCfw clDFWv"
    id="testForm"
  >
    <div
      class="sc-iqHYGH jWqqni"
    >
      <label
        for="baseField"
      >
        Test
      </label>
      <div
        class="sc-iJuUWI cFUPqY"
      >
        <input
          class="sc-fodVxV kubVCF"
          id="baseField"
          label="Test"
          name="baseField"
          value=""
        />
      </div>
    </div>
    <button
      class="sc-iBPRYJ hcYpVh"
      type="submit"
    >
      Wyślij
    </button>
  </form>
</div>
`;

exports[`base <BaseRangeField /> matches snapshot 1`] = `
<input
  class="sc-bkzZxe hpQztk"
  id="baseRangeField"
  max="10000"
  min="300"
  name="baseRangeField"
  step="100"
  value="2000"
/>
`;

exports[`base <BaseRangeField /> matches snapshot with label 1`] = `
<div>
  <form
    action="#"
    class="sc-fubCfw clDFWv"
    id="testForm"
  >
    <div
      class="sc-iqHYGH jWqqni"
    >
      <div
        class="sc-crrsfI ieBApp"
      >
        <label
          for="baseRangeField"
        >
          Test
        </label>
        <div
          class="sc-fFubgz fJyOdk"
        >
          <input
            class="sc-bkzZxe hpQztk"
            id="baseRangeField"
            label="Test"
            max="10000"
            min="300"
            name="baseRangeField"
            step="100"
            value="2000"
          />
          <span
            class="sc-bBXqnf bwzWoF"
          >
            PLN
          </span>
        </div>
      </div>
      <div
        class="sc-lmoMRL hIjsuX"
      >
        <div
          aria-orientation="horizontal"
          aria-valuemax="10000"
          aria-valuemin="300"
          aria-valuenow="2000"
          class="rangeslider rangeslider-horizontal"
        >
          <div
            class="rangeslider__fill"
            style="width: 0px;"
          />
          <div
            class="rangeslider__handle"
            style="left: 0px;"
            tabindex="0"
          >
            <div
              class="rangeslider__handle-label"
            />
          </div>
          <ul
            class="rangeslider__labels"
          />
        </div>
      </div>
    </div>
    <button
      class="sc-iBPRYJ hcYpVh"
      type="submit"
    >
      Wyślij
    </button>
  </form>
</div>
`;

exports[`base <BaseSelectField /> matches snapshot 1`] = `
<select
  class="sc-dIUggk hhFmtN"
  id="baseSelectField"
  name="baseSelectField"
  type="text"
>
  <option
    class="sc-dmlrTW hequYp"
    disabled=""
    value="select"
  >
    wybierz 
  </option>
  <option
    value="test1"
  >
    test1
  </option>
  <option
    value="test2"
  >
    test2
  </option>
</select>
`;

exports[`base <BaseSelectField /> matches snapshot with label 1`] = `
<div>
  <form
    action="#"
    class="sc-fubCfw clDFWv"
    id="testForm"
  >
    <div
      class="sc-iqHYGH jWqqni"
    >
      <label
        for="baseSelectField"
      >
        Test
      </label>
      <div
        class="sc-idOhPF buAmNl"
      >
        <select
          class="sc-dIUggk hhFmtN"
          id="baseSelectField"
          label="Test"
          name="baseSelectField"
          type="text"
        >
          <option
            class="sc-dmlrTW hequYp"
            disabled=""
            value="select"
          >
            wybierz 
          </option>
          <option
            value="test1"
          >
            test1
          </option>
          <option
            value="test2"
          >
            test2
          </option>
        </select>
        <span
          class="sc-hHftDr jmpDvw"
        />
      </div>
    </div>
    <button
      class="sc-iBPRYJ hcYpVh"
      type="submit"
    >
      Wyślij
    </button>
  </form>
</div>
`;
